File: libbids/session.py
from pathlib import Path
from typing import TYPE_CHECKING
from clibbids import Entity

if TYPE_CHECKING:
    from .subject import Subject


class Session(Entity):
    def __init__(self, subject: "Subject", idx: int):
        """Initialize a new session for a given subject

        Parameters
        ----------
        idx : int
            The id of the session
        subject_path : Path
            Path to the subject directory that owns this sessions
        """
        super(Session, self).__init__("Session", value=idx)
        self.subject: "Subject" = subject
        self.idx: int = idx
        self.path.mkdir(exist_ok=True)

    @property
    def path(self) -> Path:
        return self.subject.path.joinpath(self.id)

    @property
    def prefix(self) -> str:
        return "_".join([self.subject.id, self.id])

File: src/session.cpp
#include "session.hpp"
#include "subject.hpp"

// Session::Session(Subject const& subject, int idx)
Session::Session(Subject subject, int idx)
    : Entity("Session", std::nullopt, idx), subject_(subject) {
  std::filesystem::create_directory(this->path());
}

std::filesystem::path Session::path() const {
  std::filesystem::path subject_path = this->subject_.path();
  return subject_path / this->id();
}

std::string Session::prefix() const {
  return this->subject_.id() + "_" + this->id();
}

Subject Session::subject(void) const { return this->subject_; }
